File: shims/dotnet-proton/src/Sender.cs
/*
 * QIT .NET Apache Qpid Proton Shim - Sender
 */

using System;
using System.Collections.Generic;
using Apache.Qpid.Proton.Buffer;
using Apache.Qpid.Proton.Client;
using Apache.Qpid.Proton.Types;
using Apache.Qpid.Proton.Types.Messaging;
using Newtonsoft.Json;

namespace Qit.Shim
{
    public static class LcgHelper
    {
        public static byte[] LcgGenerateBytes(uint seed, int size)
        {
            uint state = seed & 0x7FFFFFFF;
            var result = new byte[size];
            for (int i = 0; i < size; i++)
            {
                state = (uint)((state * 1103515245u + 12345u) & 0x7FFFFFFF);
                result[i] = (byte)((state >> 16) & 0xFF);
            }
            return result;
        }

        public static string LcgGenerateString(uint seed, int size)
        {
            var raw = LcgGenerateBytes(seed, size);
            var chars = new char[size];
            for (int i = 0; i < size; i++)
                chars[i] = (char)(32 + (raw[i] % 95));
            return new string(chars);
        }

        public static List<string> GenerateCollectionElements(uint seed, int count, int elemSize)
        {
            int total = count * elemSize;
            string full = LcgGenerateString(seed, total);
            var result = new List<string>();
            for (int i = 0; i < count; i++)
                result.Add(full.Substring(i * elemSize, elemSize));
            return result;
        }

        public static List<string> GenerateMapKeys(int count)
        {
            var keys = new List<string>();
            for (int i = 0; i < count; i++)
                keys.Add($"key_{i:D4}");
            return keys;
        }
    }

    public static class Sender
    {
        public static void Send(string broker, string queue, string type, string data, bool jmsMode = false, string headersJson = null, string propertiesJson = null, string messageHeaderJson = null)
        {
            try
            {
                var testData = JsonConvert.DeserializeObject<List<TestMessage>>(data);
                Dictionary<string, Dictionary<string, string>> headers = null;
                if (!string.IsNullOrEmpty(headersJson))
                {
                    headers = JsonConvert.DeserializeObject<Dictionary<string, Dictionary<string, string>>>(headersJson);
                }
                Dictionary<string, Dictionary<string, string>> properties = null;
                if (!string.IsNullOrEmpty(propertiesJson))
                {
                    properties = JsonConvert.DeserializeObject<Dictionary<string, Dictionary<string, string>>>(propertiesJson);
                }
                var messages = new List<MessageResult>();

                // Parse broker URL
                var brokerUri = ParseBrokerUrl(broker);

                // Create client and connect
                IClient client = IClient.Create();

                ConnectionOptions options = new ConnectionOptions
                {
                    User = "artemis",
                    Password = "artemis"
                };

                using IConnection connection = client.Connect(brokerUri.Host, brokerUri.Port, options);
                using ISender sender = connection.OpenSender(queue);

                // Send all messages
                foreach (var testMsg in testData)
                {
                    var message = IMessage<object>.Create();
                    message.MessageId = testMsg.Index.ToString();

                    if (jmsMode && type == "map")
                    {
                        var subType = testMsg.Type ?? "string";
                        var key = $"{subType}_{testMsg.Index:D3}";
                        var encodedValue = TypeCodec.Encode(subType, testMsg.Value);
                        message.Body = new Dictionary<string, object> { { key, encodedValue } };
                    }
                    else if (jmsMode && type == "list")
                    {
                        var subType = testMsg.Type ?? "string";
                        var encodedValue = TypeCodec.Encode(subType, testMsg.Value);
                        message.Body = new List<object> { encodedValue };
                    }
                    else if (type == "array" || type == "described")
                    {
                        var encoded = TypeCodec.EncodeComplex(type, testMsg.Value);
                        var advanced = (IAdvancedMessage<object>)message;
                        advanced.AddBodySection(new AmqpValue(encoded));
                    }
                    else if (TypeCodec.IsComplexType(type))
                    {
                        message.Body = TypeCodec.EncodeComplex(type, testMsg.Value);
                    }
                    else
                    {
                        message.Body = TypeCodec.Encode(type, testMsg.Value);
                    }

                    // Add JMS annotations if in JMS mode
                    if (jmsMode)
                    {
                        sbyte jmsType = GetJmsMessageType(type);
                        if (jmsType >= 0)
                        {
                            // NOTE: Key MUST be symbol, value MUST be signed byte
                            // This matches Qpid JMS Client wire format
                            message.SetAnnotation("x-opt-jms-msg-type", jmsType);
                        }
                    }

                    // Apply JMS headers
                    if (headers != null)
                    {
                        if (headers.ContainsKey("JMSCorrelationID"))
                        {
                            var h = headers["JMSCorrelationID"];
                            if (h["type"] == "string")
                            {
                                message.CorrelationId = h["value"];
                            }
                            else if (h["type"] == "bytes")
                            {
                                // .NET Proton client cannot send binary correlation IDs —
                                // neither byte[] nor IProtonBuffer is accepted by the encoder
                                Console.Error.WriteLine("Send error: .NET Proton does not support binary correlation IDs");
                                Environment.Exit(1);
                            }
                        }
                        if (headers.ContainsKey("JMSReplyTo"))
                        {
                            var h = headers["JMSReplyTo"];
                            message.ReplyTo = h["value"];
                            sbyte replyType = (sbyte)(h["type"] == "topic" ? 1 : 0);
                            message.SetAnnotation("x-opt-jms-reply-to", replyType);
                        }
                        if (headers.ContainsKey("JMSType"))
                        {
                            var h = headers["JMSType"];
                            message.Subject = h["value"];
                        }
                    }

                    // Apply JMS application properties
                    if (properties != null)
                    {
                        foreach (var kvp in properties)
                        {
                            var name = kvp.Key;
                            var prop = kvp.Value;
                            var propType = prop["type"];
                            var propValue = prop["value"];

                            object typedValue = propType switch
                            {
                                "boolean" => bool.Parse(propValue),
                                "byte" => (sbyte)Convert.ToInt32(propValue, 16),
                                "short" => unchecked((short)Convert.ToInt32(propValue, 16)),
                                "int" => unchecked((int)Convert.ToUInt32(propValue, 16)),
                                "long" => unchecked((long)Convert.ToUInt64(propValue, 16)),
                                "float" => BitConverter.Int32BitsToSingle(unchecked((int)Convert.ToUInt32(propValue, 16))),
                                "double" => BitConverter.Int64BitsToDouble(unchecked((long)Convert.ToUInt64(propValue, 16))),
                                "string" => propValue,
                                _ => propValue
                            };

                            message.SetProperty(name, typedValue);
                        }
                    }

                    // Apply AMQP Header section fields
                    if (!string.IsNullOrEmpty(messageHeaderJson))
                    {
                        var mh = JsonConvert.DeserializeObject<Dictionary<string, object>>(messageHeaderJson);
                        if (mh.ContainsKey("durable"))
                            message.Durable = Convert.ToBoolean(mh["durable"]);
                        if (mh.ContainsKey("priority"))
                            message.Priority = (byte)Convert.ToInt32(mh["priority"]);
                        if (mh.ContainsKey("ttl"))
                            message.TimeToLive = Convert.ToUInt32(mh["ttl"]);
                        if (mh.ContainsKey("first_acquirer"))
                            message.FirstAcquirer = Convert.ToBoolean(mh["first_acquirer"]);
                    }

                    sender.Send(message);

                    messages.Add(new MessageResult
                    {
                        Index = testMsg.Index,
                        Type = type,
                        Value = testMsg.Value
                    });
                }

                // Output result
                var result = new
                {
                    messages,
                    stats = new { sent = messages.Count }
                };

                Console.WriteLine(JsonConvert.SerializeObject(result, Formatting.Indented));
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Send error: {ex.Message}");
                Environment.Exit(1);
            }
        }

        public static void SendLargeContent(string broker, string queue, string contentType, int size, int seed, bool jmsMode, int elements = 0, int elementSize = 0)
        {
            try
            {
                var brokerUri = ParseBrokerUrl(broker);
                IClient client = IClient.Create();
                ConnectionOptions options = new ConnectionOptions
                {
                    User = "artemis",
                    Password = "artemis"
                };

                using IConnection connection = client.Connect(brokerUri.Host, brokerUri.Port, options);
                using ISender sender = connection.OpenSender(queue);

                var message = IMessage<object>.Create();
                sbyte jmsMsgType = -1;

                if (contentType == "binary")
                {
                    message.Body = LcgHelper.LcgGenerateBytes((uint)seed, size);
                    jmsMsgType = 3; // JMS_BYTES_MESSAGE
                }
                else if (contentType == "string")
                {
                    message.Body = LcgHelper.LcgGenerateString((uint)seed, size);
                    jmsMsgType = 5; // JMS_TEXT_MESSAGE
                }
                else if (contentType == "list")
                {
                    var elems = LcgHelper.GenerateCollectionElements((uint)seed, elements, elementSize);
                    message.Body = new List<object>(elems);
                    jmsMsgType = 4; // JMS_STREAM_MESSAGE
                }
                else if (contentType == "array")
                {
                    var elems = LcgHelper.GenerateCollectionElements((uint)seed, elements, elementSize);
                    message.Body = elems.ToArray();
                }
                else if (contentType == "map")
                {
                    var elems = LcgHelper.GenerateCollectionElements((uint)seed, elements, elementSize);
                    var keys = LcgHelper.GenerateMapKeys(elements);
                    var map = new Dictionary<string, object>();
                    for (int i = 0; i < elements; i++)
                        map[keys[i]] = elems[i];
                    message.Body = map;
                    jmsMsgType = 2; // JMS_MAP_MESSAGE
                }
                else if (contentType == "described")
                {
                    var elems = LcgHelper.GenerateCollectionElements((uint)seed, elements, elementSize);
                    // Use DescribedValue (implements IDescribedType) with a symbol descriptor
                    var described = new DescribedValue(
                        Symbol.Lookup("test.large.described"),
                        new List<object>(elems));
                    message.Body = described;
                }
                else
                {
                    Console.Error.WriteLine($"Unknown large-content type: {contentType}");
                    Environment.Exit(1);
                }

                if (jmsMode && jmsMsgType >= 0)
                    message.SetAnnotation("x-opt-jms-msg-type", jmsMsgType);

                sender.Send(message);

                if (contentType == "binary" || contentType == "string")
                {
                    Console.WriteLine(JsonConvert.SerializeObject(new { sent = true, size }));
                }
                else
                {
                    Console.WriteLine(JsonConvert.SerializeObject(new { sent = true, elements, element_size = elementSize }));
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Send error: {ex.Message}");
                Environment.Exit(1);
            }
        }

        private static (string Host, int Port) ParseBrokerUrl(string broker)
        {
            var uri = new Uri(broker.StartsWith("amqp://") ? broker : $"amqp://{broker}");
            return (uri.Host, uri.Port > 0 ? uri.Port : 5672);
        }

        private static sbyte GetJmsMessageType(string amqpType)
        {
            // JMS message type constants (from Qpid JMS Client)
            const sbyte JMS_MESSAGE = 0;        // Empty message
            const sbyte JMS_MAP_MESSAGE = 2;    // Map
            const sbyte JMS_BYTES_MESSAGE = 3;  // Binary data
            const sbyte JMS_STREAM_MESSAGE = 4; // List/stream
            const sbyte JMS_TEXT_MESSAGE = 5;   // String/text

            return amqpType switch
            {
                "string" => JMS_TEXT_MESSAGE,
                "binary" => JMS_BYTES_MESSAGE,
                "null" => JMS_MESSAGE,
                "map" => JMS_MAP_MESSAGE,
                "list" => JMS_STREAM_MESSAGE,
                _ => -1
            };
        }
    }

    public class TestMessage
    {
        [JsonProperty("index")]
        public int Index { get; set; }

        [JsonProperty("type")]
        public string Type { get; set; } = "string";

        [JsonProperty("value")]
        public object Value { get; set; }
    }

    public class MessageResult
    {
        [JsonProperty("index")]
        public int Index { get; set; }

        [JsonProperty("type")]
        public string Type { get; set; }

        [JsonProperty("value")]
        public object Value { get; set; }

        [JsonProperty("headers", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<string, object> Headers { get; set; }

        [JsonProperty("properties", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<string, object> Properties { get; set; }

        [JsonProperty("message_header", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<string, object> MessageHeader { get; set; }
    }
}

File: shims/java-protonj2/src/main/java/org/apache/qpid/qit/Sender.java
/*
 * QIT ProtonJ2 Shim - Sender
 */
package org.apache.qpid.qit;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.apache.qpid.protonj2.client.Client;
import org.apache.qpid.protonj2.client.Connection;
import org.apache.qpid.protonj2.client.ConnectionOptions;
import org.apache.qpid.protonj2.client.Message;

import java.net.URI;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class Sender {
    public static void main(String[] args) throws Exception {
        // Parse command-line arguments
        String broker = null;
        String queue = null;
        String type = null;
        String data = null;
        String headersJson = null;
        String propertiesJson = null;
        boolean jmsMode = false;
        String largeContent = null;
        int size = 0;
        int seed = 0;
        int elements = 0;
        int elementSize = 0;

        for (int i = 1; i < args.length; i++) {
            String arg = args[i];

            // Check for flags (no value)
            if ("--jms-mode".equals(arg)) {
                jmsMode = true;
                continue;
            }

            // Regular options (key-value pairs)
            if (i + 1 >= args.length) {
                System.err.println("Missing value for option: " + arg);
                System.exit(1);
            }

            String key = arg.replace("--", "");
            String value = args[i + 1];

            switch (key) {
                case "broker":
                    broker = value;
                    break;
                case "queue":
                    queue = value;
                    break;
                case "type":
                    type = value;
                    break;
                case "data":
                    data = value;
                    break;
                case "headers":
                    headersJson = value;
                    break;
                case "properties":
                    propertiesJson = value;
                    break;
                case "large-content":
                    largeContent = value;
                    break;
                case "size":
                    size = Integer.parseInt(value);
                    break;
                case "seed":
                    seed = Integer.parseInt(value);
                    break;
                case "elements":
                    elements = Integer.parseInt(value);
                    break;
                case "element-size":
                    elementSize = Integer.parseInt(value);
                    break;
            }
            i++;  // Skip the value in next iteration
        }

        // Large content send path
        if (largeContent != null) {
            if (broker == null || queue == null) {
                System.err.println("Missing required arguments");
                System.exit(1);
            }

            URI brokerUri = parseBrokerUrl(broker);
            Client client = Client.create();
            ConnectionOptions options = new ConnectionOptions();
            options.user("artemis");
            options.password("artemis");

            try (Connection connection = client.connect(brokerUri.getHost(), brokerUri.getPort(), options);
                 org.apache.qpid.protonj2.client.Sender sender = connection.openSender(queue)) {

                Message<Object> message = Message.create();
                byte jmsMsgType = -1;

                if ("binary".equals(largeContent)) {
                    message.body(lcgGenerateBytes(seed, size));
                    jmsMsgType = 3;
                } else if ("string".equals(largeContent)) {
                    message.body(lcgGenerateString(seed, size));
                    jmsMsgType = 5;
                } else if ("list".equals(largeContent)) {
                    java.util.List<String> elems = generateCollectionElements(seed, elements, elementSize);
                    message.body(new java.util.ArrayList<Object>(elems));
                    jmsMsgType = 4; // STREAM_MESSAGE
                } else if ("array".equals(largeContent)) {
                    java.util.List<String> elems = generateCollectionElements(seed, elements, elementSize);
                    message.body(elems.toArray(new String[0]));
                } else if ("map".equals(largeContent)) {
                    java.util.List<String> elems = generateCollectionElements(seed, elements, elementSize);
                    java.util.List<String> keys = generateMapKeys(elements);
                    java.util.Map<String, Object> map = new java.util.LinkedHashMap<>();
                    for (int idx = 0; idx < elements; idx++) {
                        map.put(keys.get(idx), elems.get(idx));
                    }
                    message.body(map);
                    jmsMsgType = 2; // MAP_MESSAGE
                } else if ("described".equals(largeContent)) {
                    java.util.List<String> elems = generateCollectionElements(seed, elements, elementSize);
                    message.body(new org.apache.qpid.protonj2.types.UnknownDescribedType(
                        org.apache.qpid.protonj2.types.Symbol.valueOf("test.large.described"),
                        new java.util.ArrayList<Object>(elems)));
                } else {
                    System.err.println("Unknown large-content type: " + largeContent);
                    System.exit(1);
                }

                if (jmsMode && jmsMsgType >= 0) {
                    message.annotation("x-opt-jms-msg-type", jmsMsgType);
                }

                sender.send(message);
            }

            // Output result
            StringBuilder sb = new StringBuilder();
            sb.append("{\"sent\": true");
            if ("binary".equals(largeContent) || "string".equals(largeContent)) {
                sb.append(", \"size\": ").append(size);
            } else {
                sb.append(", \"elements\": ").append(elements);
                sb.append(", \"element_size\": ").append(elementSize);
            }
            sb.append("}");
            System.out.println(sb.toString());
            return;
        }

        if (broker == null || queue == null || type == null || data == null) {
            System.err.println("Missing required arguments");
            System.exit(1);
        }

        // Parse broker URL
        URI brokerUri = parseBrokerUrl(broker);

        // Parse test data
        Gson gson = new Gson();
        JsonArray testData = gson.fromJson(data, JsonArray.class);

        // Create client and connect
        Client client = Client.create();
        ConnectionOptions options = new ConnectionOptions();
        options.user("artemis");
        options.password("artemis");

        try (Connection connection = client.connect(brokerUri.getHost(), brokerUri.getPort(), options);
             org.apache.qpid.protonj2.client.Sender sender = connection.openSender(queue)) {

            JsonObject headers = null;
            if (headersJson != null) {
                headers = gson.fromJson(headersJson, JsonObject.class);
            }

            JsonObject properties = null;
            if (propertiesJson != null) {
                properties = gson.fromJson(propertiesJson, JsonObject.class);
            }

            List<JsonObject> messages = new ArrayList<>();

            // Send all messages
            for (int i = 0; i < testData.size(); i++) {
                JsonObject testMsg = testData.get(i).getAsJsonObject();
                int index = testMsg.get("index").getAsInt();
                Object value = testMsg.get("value");

                Message<Object> message = Message.create();
                message.messageId(String.valueOf(index));

                if (jmsMode && type.equals("map")) {
                    String subType = testMsg.has("type") ? testMsg.get("type").getAsString() : "string";
                    String key = String.format("%s_%03d", subType, index);
                    Object encodedValue = TypeCodec.encode(subType, value);
                    Map<String, Object> mapBody = new LinkedHashMap<>();
                    mapBody.put(key, encodedValue);
                    message.body(mapBody);
                } else if (jmsMode && type.equals("list")) {
                    String subType = testMsg.has("type") ? testMsg.get("type").getAsString() : "string";
                    Object encodedValue = TypeCodec.encode(subType, value);
                    List<Object> listBody = new ArrayList<>();
                    listBody.add(encodedValue);
                    message.body(listBody);
                } else if (TypeCodec.isComplexType(type)) {
                    message.body(TypeCodec.encodeComplex(type, value));
                } else {
                    message.body(TypeCodec.encode(type, value));
                }

                // Add JMS annotations if in JMS mode
                if (jmsMode) {
                    byte jmsType = getJmsMessageType(type);
                    if (jmsType >= 0) {
                        // NOTE: Key MUST be Symbol, value MUST be signed byte
                        // This matches Qpid JMS Client wire format
                        message.annotation("x-opt-jms-msg-type", jmsType);
                    }
                }

                // Apply JMS headers
                if (headers != null) {
                    if (headers.has("JMSCorrelationID")) {
                        JsonObject h = headers.getAsJsonObject("JMSCorrelationID");
                        String htype = h.get("type").getAsString();
                        if ("string".equals(htype)) {
                            message.correlationId(h.get("value").getAsString());
                        } else if ("bytes".equals(htype)) {
                            System.err.println("Error: ProtonJ2 does not support binary correlation IDs");
                            System.exit(1);
                        }
                    }
                    if (headers.has("JMSReplyTo")) {
                        JsonObject h = headers.getAsJsonObject("JMSReplyTo");
                        message.replyTo(h.get("value").getAsString());
                        byte replyType = (byte) ("topic".equals(h.get("type").getAsString()) ? 1 : 0);
                        message.annotation("x-opt-jms-reply-to", replyType);
                    }
                    if (headers.has("JMSType")) {
                        JsonObject h = headers.getAsJsonObject("JMSType");
                        message.subject(h.get("value").getAsString());
                    }
                }

                // Apply JMS application properties
                if (properties != null) {
                    for (Map.Entry<String, JsonElement> entry : properties.entrySet()) {
                        String name = entry.getKey();
                        JsonObject prop = entry.getValue().getAsJsonObject();
                        String ptype = prop.get("type").getAsString();
                        switch (ptype) {
                            case "boolean":
                                message.property(name, prop.get("value").getAsBoolean());
                                break;
                            case "byte": {
                                String hex = prop.get("value").getAsString().replace("0x", "");
                                long val = Long.parseUnsignedLong(hex, 16);
                                message.property(name, (byte) val);
                                break;
                            }
                            case "short": {
                                String hex = prop.get("value").getAsString().replace("0x", "");
                                long val = Long.parseUnsignedLong(hex, 16);
                                message.property(name, (short) val);
                                break;
                            }
                            case "int": {
                                String hex = prop.get("value").getAsString().replace("0x", "");
                                long val = Long.parseUnsignedLong(hex, 16);
                                message.property(name, (int) val);
                                break;
                            }
                            case "long": {
                                String hex = prop.get("value").getAsString().replace("0x", "");
                                long val = Long.parseUnsignedLong(hex, 16);
                                message.property(name, val);
                                break;
                            }
                            case "float": {
                                String hex = prop.get("value").getAsString().replace("0x", "");
                                long val = Long.parseUnsignedLong(hex, 16);
                                float floatVal = Float.intBitsToFloat((int) val);
                                message.property(name, floatVal);
                                break;
                            }
                            case "double": {
                                String hex = prop.get("value").getAsString().replace("0x", "");
                                long val = Long.parseUnsignedLong(hex, 16);
                                double doubleVal = Double.longBitsToDouble(val);
                                message.property(name, doubleVal);
                                break;
                            }
                            case "string":
                                message.property(name, prop.get("value").getAsString());
                                break;
                        }
                    }
                }

                sender.send(message);

                // Record sent message
                JsonObject msgResult = new JsonObject();
                msgResult.addProperty("index", index);
                msgResult.addProperty("type", type);

                // Explicitly handle null values - must include "value" key even when null
                JsonElement valueElement = testMsg.get("value");
                if (valueElement == null || valueElement.isJsonNull()) {
                    msgResult.add("value", com.google.gson.JsonNull.INSTANCE);
                } else {
                    msgResult.add("value", valueElement);
                }
                messages.add(msgResult);
            }

            // Output result
            JsonObject result = new JsonObject();
            JsonArray messagesArray = new JsonArray();
            for (JsonObject msg : messages) {
                messagesArray.add(msg);
            }
            result.add("messages", messagesArray);

            JsonObject stats = new JsonObject();
            stats.addProperty("sent", messages.size());
            result.add("stats", stats);

            Gson prettyGson = new GsonBuilder()
                .setPrettyPrinting()
                .serializeNulls()  // Force serialization of null values
                .create();
            System.out.println(prettyGson.toJson(result));
        }
    }

    private static URI parseBrokerUrl(String broker) throws Exception {
        if (!broker.startsWith("amqp://")) {
            broker = "amqp://" + broker;
        }
        URI uri = new URI(broker);
        return uri;
    }

    private static byte[] hexToBytes(String hex) {
        int len = hex.length();
        byte[] result = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            result[i / 2] = (byte) ((Character.digit(hex.charAt(i), 16) << 4)
                                   + Character.digit(hex.charAt(i + 1), 16));
        }
        return result;
    }

    private static byte[] lcgGenerateBytes(int seed, int size) {
        int state = seed & 0x7FFFFFFF;
        byte[] result = new byte[size];
        for (int i = 0; i < size; i++) {
            state = (int)(((long)state * 1103515245L + 12345L) & 0x7FFFFFFFL);
            result[i] = (byte)((state >> 16) & 0xFF);
        }
        return result;
    }

    private static String lcgGenerateString(int seed, int size) {
        byte[] raw = lcgGenerateBytes(seed, size);
        char[] chars = new char[size];
        for (int i = 0; i < size; i++) {
            chars[i] = (char)(32 + ((raw[i] & 0xFF) % 95));
        }
        return new String(chars);
    }

    private static java.util.List<String> generateCollectionElements(int seed, int count, int elemSize) {
        int total = count * elemSize;
        String full = lcgGenerateString(seed, total);
        java.util.List<String> result = new java.util.ArrayList<>();
        for (int i = 0; i < count; i++) {
            result.add(full.substring(i * elemSize, (i + 1) * elemSize));
        }
        return result;
    }

    private static java.util.List<String> generateMapKeys(int count) {
        java.util.List<String> keys = new java.util.ArrayList<>();
        for (int i = 0; i < count; i++) {
            keys.add(String.format("key_%04d", i));
        }
        return keys;
    }

    private static byte getJmsMessageType(String amqpType) {
        // JMS message type constants (from Qpid JMS Client)
        final byte JMS_MESSAGE = 0;        // Empty message
        final byte JMS_MAP_MESSAGE = 2;    // Map
        final byte JMS_BYTES_MESSAGE = 3;  // Binary data
        final byte JMS_STREAM_MESSAGE = 4; // List/stream
        final byte JMS_TEXT_MESSAGE = 5;   // String/text

        switch (amqpType) {
            case "string":
                return JMS_TEXT_MESSAGE;
            case "binary":
                return JMS_BYTES_MESSAGE;
            case "null":
                return JMS_MESSAGE;
            case "map":
                return JMS_MAP_MESSAGE;
            case "list":
                return JMS_STREAM_MESSAGE;
            default:
                return -1;
        }
    }
}
